### ISSUE TITLE
MaterialAddinsBase: Load external javascript files only when they are really needed.

### ISSUE THREAD

**Issue by gilberto-torrezan** [GwtMaterialDesign/gwt-material-addins#25]
Currently the `MaterialAddinsBase` class imports 9 javascript files:

```java
public void load() {
        inject(MaterialAddinsResources.INSTANCE.timepickerJs());
        inject(MaterialAddinsResources.INSTANCE.subHeaderJs());
        inject(MaterialAddinsResources.INSTANCE.pathAnimatorJs());
        inject(MaterialAddinsResources.INSTANCE.waterfallJs());
        inject(MaterialAddinsResources.INSTANCE.scrollFireJs());
        inject(MaterialAddinsResources.INSTANCE.bubbleJs());
        inject(MaterialAddinsResources.INSTANCE.interactJs());
        inject(MaterialAddinsResources.INSTANCE.masonryJs());
        inject(MaterialAddinsResources.INSTANCE.tinymceJs());
}
```
... but, if in my application, I don't use ScrollFire or Waterfall for example, I still import their scripts.

In my opinion the Widgets that use specific js files should be responsible to import them. Of course the framework can provide some sort of "JS Registry" to keep track of what is imported and what is not.

Making the components import their JS files also helps the code splitter to lazy load the files only when really needed, improving the startup time of the app.

**Comment by kevzlou7979** [comment]
Ok yeah this is true, we will separate each module resources so that the developer will just import the necessary modules they wanted to. <inherits name="gwt.material.design.addins.bubble.Bubble" /> will just use the bubble resources.

**Comment by gilberto-torrezan** [comment]
What I mean is to make the components inject their own scripts.

Something like:

```java
public class MaterialTimePicker extends MaterialWidget //...
{

    //the static initializer is called only at the first time the class is loaded
    static {
        MaterialScriptInjector.inject(MaterialAddinsResources.INSTANCE.timepickerJs());
    }
//...
```
... and create a MaterialScriptInjector to handle the inject itself (sample code):

```java
public class MaterialScriptInjector {
    public static void inject(TextResource resource) {
        inject(resource, true, false);
    }

    public static void injectDebug(TextResource resource) {
        inject(resource, false, true);
    }

    public static void inject(TextResource resource, boolean removeTag, boolean sourceUrl) {
        String text = resource.getText() +
                (sourceUrl ? "//# sourceURL="+resource.getName()+".js" : "");

        // Inject the script resource
        ScriptInjector.fromString(text)
                .setWindow(ScriptInjector.TOP_WINDOW)
                .setRemoveTag(removeTag)
                .inject();
    }
}
```
The MaterialScriptInjector should be responsible to avoid the same script to be injected more than once. The debug option could be System property (so can be toggled on and off without changing the module descriptor).

Also, with that approach, the developer doesn't need to know which modules he should import for a given widget to work. Just use the widget, and the script will be auto injected when (and only when) needed.

**Comment by BenDol** [comment]
I think a better approach for general development is to separate each component into their own "module". This way we are able to have a clear idea of each feature and what resources they use. Also means contributions are easier to test and manage.

**Comment by gilberto-torrezan** [comment]
Do you mean a GWT module? So each Widget would be needed to be declared at the `gwt.xml` ?

**Comment by BenDol** [comment]
Yeah, so for each external library wrapper it has its own GWT module that you import when you want to use it. This can of course use the same approach you are suggesting with lazy loading the components resources when able to do so. So then when we have new wrapper ports in gwt material addins their resources are easily defined in their own package.

**Comment by gilberto-torrezan** [comment]
I don't quite agree with that, having tons of module imports on `gwt.xml` doesn't help the end developer:

* It's hard to prototype: when you just want to test a widget on your layout, widgets won't "just work", you'll have to stop your super dev mode code server (yes, you need to restart it when you change the `gwt.xml` file), import the desired module (which includes searching on the docs for the right name of the module) and then test... And chances are that you forget that module on your `gwt.xml` file after the testing.
* Doesn't help the code splitter: any code referenced at the `onModuleLoad` will be there at the startup time of the application, even if you don't use that widget on the first view. That increases the first load time.
* Can create dependency loops: if one widget from one module have dependency from other modules.

I agree that the code of the widgets should be more manageable, but I don't think that creating a module for each Widget is the best solution.

**Comment by BenDol** [comment]
Creating a new addin will simply require a new xml file and an entrypoint. This means that the client path is all that needs to be defined then you are good to go. I think this is important to separate when it comes to multiple different external js library wrapping. There are use cases where static resource imports are required for certain modules. Which means there needs to be its own onModuleLoad. Otherwise the widget itself will lazy load its resources still. At this stage I think this is the best idea we have to make sure you have full control. Can see a similar implementation used by GwtBootstrap extras. https://github.com/gwtbootstrap3/gwtbootstrap3-extras/tree/master/src/main/java/org/gwtbootstrap3/extras

**Comment by BenDol** [comment]
That being said if we can make it work in an efficient way without GWT modules I'd be open to see how it would be designed (package structure, loading, etc).

**Comment by gilberto-torrezan** [comment]
> There are use cases where static resource imports are required for certain modules. Which means there needs to be its own onModuleLoad.

That's the point I'm trying to say: there's no need to create a module for each Widget if the Widget knows how to initialize itself (at my example, I used the static initializer to import the Javascript file).

Instead of a new module for each Widget, I suggest a package structure that isolates all the resources of the Widget. Taking the TimePicker as example again:

* gwt.material.design.addins.client
  * timepicker
    * MaterialTimePicker
    * resources
        * MaterialTimePickerClientBundle
        * js
          * timepicker.min.js

In other words: the `timepicker` package would contain the widget, the client bundle and the related resources. The MaterialTimePicker would load the MaterialTimePickerClientBundle at the static initialization time (avoiding it to load at the `onModuleLoad` time) and everything would be self-contained.

That same idea could apply to events and handlers associated to the widget as well.

Since everything would be self-contained at the package, transforming it in the future on a new module (if really needed), or even a separate project, would be pretty simple. One step at a time.

**Comment by BenDol** [comment]
Sure if that method will cover all the use cases it will be fine :)

**Comment by kevzlou7979** [comment]
@gilberto-torrezan  and @BenDol , I have started to develop the reorganization of the Addins widget with the gwtbootsrap-extras. With exact packaging convention that have been done on bootstrap extras.

**Comment by BenDol** [comment]
Do as @gilberto-torrezan suggests, we wont use GWT modules for each external library wrapper. Still packaging like the gwtbootstrap extras but just without its own entrypoint/gwt.xml file. Just read his suggestion.

**Comment by kevzlou7979** [comment]
Ok got it

**Comment by kevzlou7979** [comment]
I got the POC work , the idea of @gilberto-torrezan and the packaging of bootstrap-extras will submit a pr or create a branch soon

**Comment by kevzlou7979** [comment]
@gilberto-torrezan and @BenDol  I have create a new branch for the said implementation take a look, I have tested some of it and it works well https://github.com/GwtMaterialDesign/gwt-material-addins/tree/resource_reorganization

**Comment by kevzlou7979** [comment]
Can we put this update on another version like 1.5.1 because some users already using this as snapshot version. Some are asking if we can put this on another version. What do you think @gilberto-torrezan  and @BenDol

**Comment by gilberto-torrezan** [comment]
I'd call that a `1.6.0`, because it breaks existing code. I don't know how are you versioning the project, but I like the way Apache do: `major.minor.bugfix`.

**Comment by kevzlou7979** [comment]
@gilberto-torrezan Yeah ok I agree, lets make it 1.6.0.

**Comment by kevzlou7979** [comment]
So everytime we do some changes on the master (1.5.0 SNAPSHOT) will be automatically updated the 1.6.0

**Comment by BenDol** [comment]
Well 1.5.0-SNAPSHOT is expected to break, we haven't even released 1.5.0 release version yet. We should actually keep it as 1.5.0-SNAPSHOT and the people using the snapshot version need to update.

**Comment by kevzlou7979** [comment]
Ok, this was implemented to all addin widgets.

**Comment by BenDol** [comment]
There is an issue we have run into with this technique you have suggested, Gilberto. We need to be able to provide public resources for the rich editor addin. TinyMCE requires some skin/template resources that are loaded from the tinymce.js file. Since we don't have a gwt module for each feature we can't include these public resource now. Do you have a solution for this?

**Comment by gilberto-torrezan** [comment]
@BenDol Well, TinyMCE is no ordinary component, it has tons of features, modules and extensions, and the community version is licensed under LGPL 2.1 (while GWT Material uses Apache License 2.0). 

Because of how complex and different it is from the rest of the other widgets of the addins project, I suggest it to be treated like the Material Charts is: a separate project, with a different jar (which means: with its own module and public folder).

But, abstracting the TinyMCE component, let's suppose there's another component we would like to add to the addins project, which it is pretty simple and still requires a public folder. In this case I see no other option other than creating a module for it: the package structure can be the same, the lazy loading of scripts (on static blocks) can be the same, but a module would be created.

(Well, actually another option would be to create a single public folder for all components, but that would defeat the purpose of the isolation of the components).

Still, in my opinion, having some modules that doesn't mean that *every* component needs its own module. The majority of the components don't need a public folder.

**Comment by BenDol** [comment]
I don't think we should separate the rich editor out into another jar since its just a standard addin. We can create a GWT module for it however and make sure we document it for users. That way we can define the public resources required. 

But to be honest its not uncommon for a library to require public resources like fonts, images, etc so might end up having to use GWT modules for quite a few addins anyway. Which is why I was opting for that approach in the first place, so we don't have some addins that need to be imported and others that don't.

But anyway this should work fine for now, so long as its documented well it will be fine.

**Comment by kevzlou7979** [comment]
Ok so we will go for Module based wysiwyg will develop this now and add a
note in the RichEditor page showcase

**Comment by gilberto-torrezan** [comment]
What about the [license issue](http://www.apache.org/legal/resolved.html#category-x)?

**Comment by BenDol** [comment]
Yea will need to look into that. Another Rich Editor might need to be used in this case.

**Comment by kevzlou7979** [comment]
I found one that suits gwt-material https://github.com/Cerealkillerway/materialNote

**Comment by kevzlou7979** [comment]
It's lightweight and MIT License will implement this now, I created a POC about this thing

**Comment by BenDol** [comment]
Looks good :+1:

**Comment by kevzlou7979** [comment]
Yeah will do some revisions into , to handle more professional than it was.
